rtl/eclrtl/rtlrecord.cpp

```diff
@@ -260,3 +352,132 @@ RtlDynRow::~RtlDynRow()
     delete [] variableOffsets;
 }
 
+//-----------------
+
+const RtlRecord *setupRecordAccessor(const COutputMetaData &meta, bool expand, std::atomic<const RtlRecord *> &aRecordAccessor)
+{
+    const RtlRecord *lRecordAccessor = new RtlRecord(meta.queryTypeInfo()->queryFields(), expand);
+    const RtlRecord *expected = nullptr;
+    if (aRecordAccessor.compare_exchange_strong(expected, lRecordAccessor))
+        return lRecordAccessor;
+    else
+    {
+        // Other thread already set it while we were creating
+        delete lRecordAccessor;
+        return expected;  // has been updated to the value set by other thread
+    }
+}
+
+const RtlRecord *COutputMetaData::queryRecordAccessor(bool expand) const
+{
+    // NOTE: the recordAccessor field cannot be declared as atomic, since the class definition is included in generated
+    // code which is not (yet) compiled using C++11. If that changes then the reinterpret_cast can be removed.
+    std::atomic<const RtlRecord *> &aRecordAccessor = reinterpret_cast<std::atomic<const RtlRecord *> &>(recordAccessor[expand]);
```

Comment:
I wondered if index[bool] was safe, but it looks like it is - explicitly defined to cast fasle->0 true->1